src/devices/src/virtio/net/simple_handler.rs

```diff
@@ -0,0 +1,188 @@
+// as one output descriptor to the transmitq, and the device is notified of the new entry".
+// We assume the TX frame will not exceed this size either.
+const MAX_BUFFER_SIZE: usize = 65562;
+
+#[derive(Debug)]
+pub enum Error {
+    GuestMemory(vm_memory::GuestMemoryError),
+    Queue(vm_virtio::Error),
+    Tap(io::Error),
+}
+
+impl From<vm_virtio::Error> for Error {
+    fn from(e: vm_virtio::Error) -> Self {
+        Error::Queue(e)
+    }
+}
+
+// A simple handler implementation for a RX/TX queue pair, which does not make assumptions about
+// the way queue notification is implemented. The backend is not yet generic (we always assume a
+// `Tap` object), but we're looking at improving that going forward.
+// TODO: Find a better name.
+pub struct SimpleHandler<M: GuestAddressSpace, S: SignalUsedQueue> {
+    pub driver_notify: S,
+    pub rxq: Queue<M>,
+    pub rxbuf_current: usize,
+    pub rxbuf: [u8; MAX_BUFFER_SIZE],
+    pub txq: Queue<M>,
+    pub txbuf: [u8; MAX_BUFFER_SIZE],
+    pub tap: Tap,
+}
+
+impl<M: GuestAddressSpace, S: SignalUsedQueue> SimpleHandler<M, S> {
+    pub fn new(driver_notify: S, rxq: Queue<M>, txq: Queue<M>, tap: Tap) -> Self {
+        SimpleHandler {
+            driver_notify,
+            rxq,
+            rxbuf_current: 0,
+            rxbuf: [0u8; MAX_BUFFER_SIZE],
+            txq,
+            txbuf: [0u8; MAX_BUFFER_SIZE],
+            tap,
+        }
+    }
+
+    // Have to see how to approach error handling for the `Queue` implementation in particular,
+    // because many situations are not really recoverable. We should consider reporting them based
+    // on the  metrics/events solution when they appear, and not propagate them further unless
+    // it's really useful/necessary.
+    fn write_frame_to_guest(&mut self) -> result::Result<bool, Error> {
+        let num_bytes = self.rxbuf_current;
+
+        let mut chain = match self.rxq.iter()?.next() {
+            Some(c) => c,
+            _ => return Ok(false),
+        };
+
+        let mut count = 0;
+        let buf = &mut self.rxbuf[..num_bytes];
+
+        while let Some(desc) = chain.next() {
+            let left = buf.len() - count;
+
+            if left == 0 {
+                break;
+            }
+
+            let len = cmp::min(left, desc.len() as usize);
+            chain
+                .memory()
+                .write_slice(&buf[count..count + len], desc.addr())
+                .map_err(Error::GuestMemory)?;
+
+            count += len;
+        }
+
+        if count != buf.len() {
+            // The frame was too large for the chain.
+            warn!("rx frame too large");
+        }
+
+        self.rxq.add_used(chain.head_index(), count as u32)?;
+
+        self.rxbuf_current = 0;
+
+        Ok(true)
+    }
+
+    pub fn process_tap(&mut self) -> result::Result<(), Error> {
+        loop {
+            if self.rxbuf_current == 0 {
+                match self.tap.read(&mut self.rxbuf) {
```

Comment:
I'm not quite sure what happens if we read some data from tap here, but don't have a descriptor chain available for writing to its buffer. Is that data lost? Should we do this `read` only if we're sure there is at least one descriptor chain?